fix(cody): use client-provided timeout for completions requests

taras-yemets opened the pull request:
Closes [CODY-2775](https://linear.app/sourcegraph/issue/CODY-2775/%5Bautocomplete-latency%5D-apply-the-same-timeout-on-the-cody-gateway-side)



Enables client control over the request processing timeout on the server (both Sourcegraph backend and Cody Gateway). The context timeout is set to the value provided in the `X-Timeout-Ms` header of the client request. If the header is not provided, the default context timeout is used (1 minute on both Sourcegraph backend and Cody Gateway).

Previously, we only had a default timeout on the Sourcegraph backend side (8 minutes).

Corresponding client change:
- https://github.com/sourcegraph/cody/pull/4921

<!-- 💡 To write a useful PR description, make sure that your description covers:
- WHAT this PR is changing:
    - How was it PREVIOUSLY.
    - How it will be from NOW on.
- WHY this PR is needed.
- CONTEXT, i.e. to which initiative, project or RFC it belongs.

The structure of the description doesn't matter as much as covering these points, so use
your best judgement based on your context.
Learn how to write good pull request description: https://www.notion.so/sourcegraph/Write-a-good-pull-request-description-610a7fd3e613496eb76f450db5a49b6e?pvs=4 -->


## Test plan
- Manually tested and confirmed that if the request contains the `X-Timeout-Ms` header, its value is used. If not, the default maximum request duration is applied.
- CI
- 
<!-- All pull requests REQUIRE a test plan: https://docs-legacy.sourcegraph.com/dev/background-information/testing_principles -->


## Changelog
- Use the provided timeout from request parameters if available; otherwise use the default maximum request duration (8 minutes)

<!--
1. Ensure your pull request title is formatted as: $type($domain): $what
2. Add bullet list items for each additional detail you want to cover (see example below)
3. You can edit this after the pull request was merged, as long as release shipping it hasn't been promoted to the public.
4. For more information, please see this how-to https://www.notion.so/sourcegraph/Writing-a-changelog-entry-dd997f411d524caabf0d8d38a24a878c?

Audience: TS/CSE > Customers > Teammates (in that order).

Cheat sheet: $type = chore|fix|feat $domain: source|search|ci|release|plg|cody|local|...
-->

<!--
Example:

Title: fix(search): parse quotes with the appropriate context
Changelog section:

## Changelog

- When a quote is used with regexp pattern type, then ...
- Refactored underlying code.
-->

taras-yemets commented on cmd/frontend/internal/httpapi/completions/handler.go:
@rafax, is it the correct place to perform this check? Do we need to apply this check for all requests or only for those that are [going to be sent to the Cody Gateway](https://github.com/sourcegraph/sourcegraph/blob/13b134eef525edbc757e31afa96b6d7954e7b283/cmd/frontend/internal/httpapi/completions/handler.go#L148-L152)?

rafax replied:
I think this is the correct place to perform this check _in the Sourcegraph backend_ (where it will apply to Cody Gateway and BYOK/BYOLLM deployments). I'm not 100% sure if SG is responsible for the cancellation propagation bugs, so we may need to do the same thing in Cody Gateway.

taras-yemets replied:
Completion client's [`Stream`](https://github.com/sourcegraph/sourcegraph/blob/13b134eef525edbc757e31afa96b6d7954e7b283/cmd/frontend/internal/httpapi/completions/handler.go#L390)/[`Complete`](https://github.com/sourcegraph/sourcegraph/blob/13b134eef525edbc757e31afa96b6d7954e7b283/cmd/frontend/internal/httpapi/completions/handler.go#L464) methods are called with this context (with timeout). So I expect the context cancelation to be handled by the corresponding method of a matching client returned by [`codyGatewayClient.clientForParams`](https://github.com/sourcegraph/sourcegraph/blob/13b134eef525edbc757e31afa96b6d7954e7b283/internal/completions/client/codygateway/codygateway.go#L80-L159).
I'm new to this part of the codebase, so this may be a wrong assumption. WDYT, @rafax?

chrsmith replied:
IMO, we should apply this check for all requests, regardless of the completions provider it is sent to. It would be super-lame if the `X-Timeout-Ms` HTTP header was only honored by Cody Gateway, and completely ignored if the Sourcegraph backend was configured any other way.

taras-yemets replied:
This PR implements the same `X-Tineout-Ms` header checks both on the Sourcegraph backend and on the Cody Gateway.

valerybugakov commented on cmd/frontend/internal/httpapi/completions/handler.go:
```diff
@@ -67,7 +67,24 @@ func newCompletionsHandler(
+		} else {
+			ctxTimeout = maxRequestDuration
+		}
+		ctx, cancel := context.WithTimeout(r.Context(), ctxTimeout)
```
I believe we need to add similar logic on the Cody Gateway side to make it work with fast-path, where requests from IDE extensions go directly to Cody Gateway and bypass the `frontend` service entirely. 

https://github.com/sourcegraph/sourcegraph/blob/dfb55a275e0b7a0f48bcc15f93311dd9ce7d294f/cmd/cody-gateway/internal/httpapi/completions/upstream.go#L299-L300

valerybugakov reviewed (commented):
The fast path support for this task is critical because most of the autocomplete traffic goes directly to Cody Gateway.

taras-yemets commented:
@rafax, @valerybugakov, I updated the PR to use the timeout value provided via the `X-Timeout-Ms` header. 
I also added a default maximum duration of 5 minutes on the Gateway side. This value is arbitrary; I'm not sure why we have an 8-minute timeout on the Sourcegraph backend, so I chose a slightly smaller value. 
Can you please review the PR again?

rafax commented on cmd/cody-gateway/internal/httpapi/completions/upstream.go:
```diff
@@ -116,6 +116,10 @@ type UpstreamRequest interface {
 
+// maxRequestDuration is the maximum amount of time a request can take before
+// being cancelled as DeadlineExceeded.
+const maxRequestDuration = 5 * time.Minute
```
Link to https://github.com/sourcegraph/infrastructure/blob/a697a5c2f3f01fc0886f4c6b6efec2ae5f40f1cb/cody-gateway/modules/cloudrun/cloudrun.tf#L67 that has some extra context?

rafax replied:
Overall, it pains me to see the default timeout so high - we should really default to a shorter timeout (10s) and make long timeouts opt-in, but I know it's not in scope of this change.

taras-yemets replied:
We may want timeout per feature (chat/completions). WDYT, @chrsmith?

taras-yemets replied:
Reduced the timeout to 1 minute for Cody Gateway.
8 minutes timeout remains unchanged for the Sourcegraph backend.

taras-yemets replied:
Reduced the max request duration to 1 minute on the Sourcegraph backend too.

rafax commented on cmd/cody-gateway/internal/httpapi/completions/upstream.go:
```diff
@@ -220,6 +235,9 @@ func makeUpstreamHandler[ReqT UpstreamRequest](
 			response.JSONError(logger, w, http.StatusBadRequest, err)
 			return
 		}
+		// fmt.Println()
```
clean this up?

rafax commented on cmd/cody-gateway/internal/httpapi/completions/upstream.go:
```diff
@@ -395,7 +413,15 @@ func makeUpstreamHandler[ReqT UpstreamRequest](
 		}()
+		// fmt.Println()
+		// fmt.Printf("CODY GATEWAY HANDLER: %+v\n", upstreamRequest)
+		// fmt.Println()
```
and here

rafax commented on cmd/cody-gateway/internal/httpapi/completions/upstream.go:
```diff
@@ -395,7 +413,15 @@ func makeUpstreamHandler[ReqT UpstreamRequest](
+		// fmt.Println()
+
 		resp, err := httpClient.Do(upstreamRequest)
+		fmt.Println()
```
and here?

rafax reviewed (approved):
This LGTM, but I think you really want @slimsag / @chrsmith 👁️ on this

chrsmith commented on cmd/cody-gateway/internal/httpapi/completions/upstream.go:
```diff
@@ -161,7 +166,18 @@ func makeUpstreamHandler[ReqT UpstreamRequest](
+		ctxTimeout := maxRequestDuration
+		if v := downstreamRequest.Header.Get("X-Timeout-Ms"); v != "" {
+			if t, err := strconv.Atoi(v); err == nil {
+				ctxTimeout = time.Duration(t) * time.Millisecond
```
Nit: I can totally see us scratching our head as to why the `X-Timeout-Ms` header is getting ignored. So I would suggest we log a warning if the call to `Atoi` fails. e.g.

```go
if t, err := strong.Atoi(v); err != nil {
    logger.Warn("error parsing X-Timeout-Ms header", log.Error(err))
} else {
    ctxTimeout = time.Duration(t) * time.Millisecond
}
```

chrsmith reviewed (approved):
Minor nits, suggestions, and comments. But this all looks good to me as well.
